File: Serene/Serene.Script/Administration/UserPermission/PermissionCheckEditor.cs
using jQueryApi;
using Serenity;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;

namespace Serene.Administration
{
    [Editor, IdProperty("Key")]
    public class PermissionCheckEditor : DataGrid<PermissionCheckItem, PermissionCheckEditorOptions>
    {
        private string containsText;
        private ILookup<string, PermissionCheckItem> byParentKey;
        private JsDictionary<string, bool> rolePermissions;

        public PermissionCheckEditor(jQueryObject div, PermissionCheckEditorOptions opt)
            : base(div, opt)
        {
            rolePermissions = new JsDictionary<string, bool>();

            JsDictionary<string, string> titleByKey;
            var permissionKeys = GetSortedGroupAndPermissionKeys(out titleByKey);
            var items = new List<PermissionCheckItem>();
            foreach (var key in permissionKeys)
                items.Add(new PermissionCheckItem
                {
                    Key = key,
                    ParentKey = GetParentKey(key),
                    Title = titleByKey[key],
                    GrantRevoke = null,
                    IsGroup = key.EndsWith(":")
                });

            byParentKey = items.ToLookup(x => x.ParentKey);

            SetItems(items);
        }

        private string GetItemGrantRevokeClass(PermissionCheckItem item, bool grant)
        {
            if (!item.IsGroup)
                return item.GrantRevoke == grant ? " checked" : "";

            var desc = GetDescendants(item);
            var granted = desc.Where(x => x.GrantRevoke == grant);
            if (!granted.Any())
                return "";
            else if (desc.Count() == granted.Count())
                return "checked";
            else
                return "checked partial";
        }

        private string GetItemEffectiveClass(PermissionCheckItem item)
        {
            if (item.IsGroup)
            {
                var desc = GetDescendants(item);
                var grantCount = desc.Count(x => x.GrantRevoke == true ||
                    (x.GrantRevoke == null && rolePermissions[x.Key]));

                if (grantCount == desc.Count || desc.Count == 0)
                    return "allow";
                else if (grantCount == 0)
                    return "deny";
                else
                    return "partial";
            }
            else
            {
                bool granted = item.GrantRevoke == true ||
                    (item.GrantRevoke == null && rolePermissions[item.Key]);

                return granted ? " allow" : " deny";
            }
        }

        protected override List<SlickColumn> GetColumns()
        {
            var columns = new List<SlickColumn>
            {
                new SlickColumn
                {
                    Title = Q.Text("Site.UserPermissionDialog.Permission"),
                    Field = "Title",
                    Format = SlickFormatting.TreeToggle(
                        getView: () => this.View, 
                        getId: x => x.Key, 
                        formatter: ctx => {
                            var item = (PermissionCheckItem)ctx.Item;
                            var klass = GetItemEffectiveClass(item);
                            return "<span class='effective-permission " + klass + "'>" +
                                Q.HtmlEncode(ctx.Value) + "</span>";
                        }),
                    Width = 495,
                    Sortable = false
                },
                new SlickColumn
                {
                    Title = Q.Text("Site.UserPermissionDialog.Grant"),
                    Field = "Grant",
                    Format = ctx =>
                    {
                        var item = (PermissionCheckItem)ctx.Item;
                        string klass= GetItemGrantRevokeClass(item, true);
                        return "<span class='check-box grant no-float " +  klass + "'></span>";
                    },
                    Width = 65,
                    Sortable = false,
                    HeaderCssClass = "align-center",
                    CssClass = "align-center"
                }
            };

            if (options.ShowRevoke)
            {
                columns.Add(new SlickColumn
                {
                    Title = Q.Text("Site.UserPermissionDialog.Revoke"),
                    Field = "Revoke",
                    Format = ctx =>
                    {
                        var item = (PermissionCheckItem)ctx.Item;
                        string klass = GetItemGrantRevokeClass(item, false);
                        return "<span class=\"check-box revoke no-float " + klass + "\"></span>";
                    },
                    Width = 65,
                    Sortable = false,
                    HeaderCssClass = "align-center",
                    CssClass = "align-center"
                });
            }

            return columns;
        }

        private void SetItems(List<PermissionCheckItem> items)
        {
            SlickTreeHelper.SetIndents(items,
                getId: x => x.Key,
                getParentId: x => x.ParentKey,
                setCollapsed: false);

            this.View.SetItems(items, true);
        }

        protected override bool OnViewSubmit()
        {
            return false;
        }

        protected override bool OnViewFilter(PermissionCheckItem item)
        {
            if (!base.OnViewFilter(item))
                return false;

            if (!SlickTreeHelper.FilterById(item, view, getParentId: x => x.ParentKey))
                return false;

            if (!string.IsNullOrEmpty(containsText))
            {
                return (MatchContains(item)) ||
                        (item.IsGroup && GetDescendants(item, false).Any(MatchContains));
            }

            return true;
        }

        private bool MatchContains(PermissionCheckItem item)
        {
            return Q.Externals.StripDiacritics(item.Title ?? "").ToLower()
                .IndexOf(containsText) >= 0;
        }

        private List<PermissionCheckItem> GetDescendants(PermissionCheckItem item, bool excludeGroups = true)
        {
            var result = new List<PermissionCheckItem>();
            var stack = new Stack<PermissionCheckItem>();

            stack.Push(item);

            while (stack.Count > 0)
            {
                var i = stack.Pop();

                foreach (var child in byParentKey[i.Key])
                {
                    if (!excludeGroups || !child.IsGroup)
                        result.Add(child);

                    stack.Push(child);
                }
            }

            return result;
        }

        protected override void OnClick(jQueryEvent e, int row, int cell)
        {
            base.OnClick(e, row, cell);

            if (!e.IsDefaultPrevented())
                SlickTreeHelper.ToggleClick(e, row, cell, view, getId: x => x.Key);

            if (e.IsDefaultPrevented())
                return;

            var target = J(e.Target);

            bool? grant = target.HasClass("grant");

            if (grant.Value || target.HasClass("revoke"))
            {
                e.PreventDefault();
                var item = Rows[row];
                bool checkedOrPartial = target.HasClass("checked") || target.HasClass("partial");

                if (checkedOrPartial)
                    grant = null;
                else
                    grant = grant.Value ^ checkedOrPartial;

                if (item.IsGroup)
                {
                    foreach (var d in GetDescendants(item))
                        if (d.GrantRevoke != grant)
                        {
                            d.GrantRevoke = grant;
                        }
                }
                else if (item.GrantRevoke != grant)
                {
                    item.GrantRevoke = grant;
                }

                slickGrid.Invalidate();
            }
        }

        private string GetParentKey(string key)
        {
            if (key.EndsWith(":"))
                key = key.Substr(0, key.Length - 1);

            var idx = key.LastIndexOf(':');
            if (idx >= 0)
                return key.Substr(0, idx + 1);

            return null;
        }

        protected override List<ToolButton> GetButtons()
        {
            return new List<ToolButton>();
        }

        protected override void CreateToolbarExtensions()
        {
            base.CreateToolbarExtensions();

            GridUtils.AddQuickSearchInputCustom(toolbar.Element, (field, text) =>
            {
                containsText = Q.Externals.StripDiacritics(text.TrimToNull() ?? "").ToLower();
                view.SetItems(view.GetItems(), true);
            });
        }

        private List<string> GetSortedGroupAndPermissionKeys(out JsDictionary<string, string> titleByKey)
        {
            var keys = Q.GetRemoteData<ListResponse<string>>("Administration.PermissionKeys").Entities;

            titleByKey = new JsDictionary<string, string>();
            var titleWithGroup = new JsDictionary<string, string>();

            foreach (var k in keys)
            {
                var s = k;
                if (string.IsNullOrEmpty(s))
                    continue;

                if (s.EndsWith(":")) // permission keys can't end with ':'
                {
                    s = s.Substr(0, s.Length - 1);
                    if (s.Length == 0)
                        continue;
                }

                if (titleByKey.ContainsKey(s))
                    continue;

                titleByKey[s] = Q.TryGetText("Permission." + s) ?? s;

                var parts = s.Split(':');
                var group = "";
                var groupTitle = "";
                for (var i = 0; i < parts.Length - 1; i++)
                {
                    group = group + parts[i] + ":";
                    titleByKey[group] = Q.TryGetText("Permission." + group) ?? parts[i];
                    groupTitle = groupTitle + titleByKey[group] + ":";
                    titleWithGroup[group] = groupTitle;
                }

                titleWithGroup[s] = groupTitle + titleByKey[s];
            }

            keys = titleByKey.Keys.ToList();
            keys.Sort((x, y) => Q.Externals.TurkishLocaleCompare(
                titleWithGroup[x], titleWithGroup[y]));

            return keys;
        }

        public List<UserPermissionRow> Value
        {
            get
            {
                var result = new List<UserPermissionRow>();

                foreach (var item in this.view.GetItems())
                    if (item.GrantRevoke != null && !item.Key.EndsWith(":"))
                        result.Add(new UserPermissionRow
                        {
                            PermissionKey = item.Key,
                            Grant = item.GrantRevoke.Value
                        });

                return result;
            }
            set
            {
                foreach (var item in this.view.GetItems())
                    item.GrantRevoke = null;

                if (value != null)
                    foreach (var row in value)
                    {
                        var r = view.GetItemById(row.PermissionKey);
                        if (r != null)
                            r.GrantRevoke = row.Grant ?? true;
                    }

                SetItems(Items);
            }
        }

        public List<string> RolePermissions
        {
            get { return rolePermissions.Keys.ToList(); }
            set
            {
                rolePermissions.Clear();
                if (value != null)
                    foreach (var k in value)
                        rolePermissions[k] = true;
                SetItems(this.Items);
            }
        }
    }

    [Imported, Serializable, PreserveMemberCase]
    public class PermissionCheckItem
    {
        public string ParentKey { get; set; }
        public string Key { get; set; }
        public string Title { get; set; }
        public bool IsGroup { get; set; }
        public bool? GrantRevoke { get; set; }
    }

    [Imported, Serializable]
    public class PermissionCheckEditorOptions
    {
        public bool ShowRevoke { get; set; }
    }
}

File: Serene/Serene.Web/Modules/Administration/UserPermission/PermissionCheckEditor.ts
namespace Serene.Administration {

    @Serenity.Decorators.registerClass([Serenity.IGetEditValue, Serenity.ISetEditValue])
    export class PermissionCheckEditor extends Serenity.DataGrid<PermissionCheckItem, PermissionCheckEditorOptions> {

        protected getIdProperty() { return "Key"; }

        private searchText: string;
        private byParentKey: Q.Grouping<PermissionCheckItem>;
        private rolePermissions: Q.Dictionary<boolean> = {};

        constructor(container: JQuery, opt: PermissionCheckEditorOptions) {
            super(container, opt);

            let titleByKey: Q.Dictionary<string> = {};
            let permissionKeys = this.getSortedGroupAndPermissionKeys(titleByKey);
            let items = permissionKeys.map(key => <PermissionCheckItem>{
                Key: key,
                ParentKey: this.getParentKey(key),
                Title: titleByKey[key],
                GrantRevoke: null,
                IsGroup: key.charAt(key.length - 1) === ':'
            });

            this.byParentKey = Q.toGrouping(items, x => x.ParentKey);
            this.setItems(items);
        }

        private getItemGrantRevokeClass(item: PermissionCheckItem, grant: boolean): string {
            if (!item.IsGroup) {
                return ((item.GrantRevoke === grant) ? ' checked' : '');
            }

            let desc = this.getDescendants(item, true);
            let granted = desc.filter(x => x.GrantRevoke === grant);

            if (!granted.length) {
                return '';
            }

            if (desc.length === granted.length) {
                return 'checked';
            }

            return 'checked partial';
        }

        private getItemEffectiveClass(item: PermissionCheckItem): string {

            if (item.IsGroup) {
                let desc = this.getDescendants(item, true);
                let grantCount = Q.count(desc, x => x.GrantRevoke === true ||
                    (x.GrantRevoke == null && this.rolePermissions[x.Key]));

                if (grantCount === desc.length || desc.length === 0) {
                    return 'allow';
                }

                if (grantCount === 0) {
                    return 'deny';
                }

                return 'partial';
            }

            let granted = item.GrantRevoke === true ||
                (item.GrantRevoke == null && this.rolePermissions[item.Key]);

            return (granted ? ' allow' : ' deny');
        }

        protected getColumns(): Slick.Column[] {
            let columns: Slick.Column[] = [{
                name: Q.text('Site.UserPermissionDialog.Permission'),
                field: 'Title',
                format: Serenity.SlickFormatting.treeToggle(() => this.view, x => x.Key, ctx => {
                    let item = ctx.item;
                    let klass = this.getItemEffectiveClass(item);
                    return '<span class="effective-permission ' + klass + '">' + Q.htmlEncode(ctx.value) + '</span>';
                }),
                width: 495,
                sortable: false
            }, {
                name: Q.text('Site.UserPermissionDialog.Grant'), field: 'Grant',
                format: ctx => {
                    let item1 = ctx.item;
                    let klass1 = this.getItemGrantRevokeClass(item1, true);
                    return "<span class='check-box grant no-float " + klass1 + "'></span>";
                },
                width: 65,
                sortable: false,
                headerCssClass: 'align-center',
                cssClass: 'align-center'
            }];

            if (this.options.showRevoke) {
                columns.push({
                    name: Q.text('Site.UserPermissionDialog.Revoke'), field: 'Revoke',
                    format: ctx => {
                        let item2 = ctx.item;
                        let klass2 = this.getItemGrantRevokeClass(item2, false);
                        return '<span class="check-box revoke no-float ' + klass2 + '"></span>';
                    },
                    width: 65,
                    sortable: false,
                    headerCssClass: 'align-center',
                    cssClass: 'align-center'
                });
            }

            return columns;
        }

        public setItems(items: PermissionCheckItem[]): void {
            Serenity.SlickTreeHelper.setIndents(items, x => x.Key, x => x.ParentKey, false);
            this.view.setItems(items, true);
        }

        protected onViewSubmit() {
            return false;
        }

        protected onViewFilter(item: PermissionCheckItem): boolean {
            if (!super.onViewFilter(item)) {
                return false;
            }

            if (!Serenity.SlickTreeHelper.filterById(item, this.view, x => x.ParentKey))
                return false;

            if (this.searchText) {
                return this.matchContains(item) || item.IsGroup && Q.any(this.getDescendants(item, false), x => this.matchContains(x));
            }

            return true;
        }

        private matchContains(item: PermissionCheckItem): boolean {
            return Select2.util.stripDiacritics(item.Title || '').toLowerCase().indexOf(this.searchText) >= 0;
        }

        private getDescendants(item: PermissionCheckItem, excludeGroups: boolean): PermissionCheckItem[] {
            let result: PermissionCheckItem[] = [];
            let stack = [item];
            while (stack.length > 0) {
                let i = stack.pop();
                let children = this.byParentKey[i.Key];
                if (!children)
                    continue;

                for (let child of children) {
                    if (!excludeGroups || !child.IsGroup) {
                        result.push(child);
                    }

                    stack.push(child);
                }
            }

            return result;
        }

        protected onClick(e, row, cell): void {
            super.onClick(e, row, cell);

            if (!e.isDefaultPrevented()) {
                Serenity.SlickTreeHelper.toggleClick(e, row, cell, this.view, x => x.Key);
            }

            if (e.isDefaultPrevented()) {
                return;
            }

            let target = $(e.target);
            let grant = target.hasClass('grant');

            if (grant || target.hasClass('revoke')) {
                e.preventDefault();

                let item = this.itemAt(row);
                let checkedOrPartial = target.hasClass('checked') || target.hasClass('partial');

                if (checkedOrPartial) {
                    grant = null;
                }
                else {
                    grant = grant !== checkedOrPartial;
                }

                if (item.IsGroup) {
                    for (var d of this.getDescendants(item, true)) {
                        d.GrantRevoke = grant;
                    }
                }
                else
                    item.GrantRevoke = grant;

                this.slickGrid.invalidate();
            }
        }

        private getParentKey(key): string {
            if (key.charAt(key.length - 1) === ':') {
                key = key.substr(0, key.length - 1);
            }

            let idx = key.lastIndexOf(':');
            if (idx >= 0) {
                return key.substr(0, idx + 1);
            }
            return null;
        }

        protected getButtons(): Serenity.ToolButton[] {
            return [];
        }

        protected createToolbarExtensions(): void {
            super.createToolbarExtensions();
            Serenity.GridUtils.addQuickSearchInputCustom(this.toolbar.element, (field, text) => {
                this.searchText = Select2.util.stripDiacritics(Q.trimToNull(text) || '').toLowerCase();
                this.view.setItems(this.view.getItems(), true);
            });
        }

        private getSortedGroupAndPermissionKeys(titleByKey: Q.Dictionary<string>): string[] {
            let keys = <string[]>Q.getRemoteData('Administration.PermissionKeys').Entities;
            let titleWithGroup = {};
            for (var k of keys) {
                let s = k;

                if (!s) {
                    continue;
                }

                if (s.charAt(s.length - 1) == ':') {
                    s = s.substr(0, s.length - 1);
                    if (s.length === 0) {
                        continue;
                    }
                }

                if (titleByKey[s]) {
                    continue;
                }

                titleByKey[s] = Q.format(Q.tryGetText('Permission.' + s), s);
                let parts = s.split(':');
                let group = '';
                let groupTitle = '';
                for (let i = 0; i < parts.length - 1; i++) {
                    group = group + parts[i] + ':';
                    let txt = Q.tryGetText('Permission.' + group);
                    if (txt == null) {
                        txt = parts[i];
                    }
                    titleByKey[group] = txt;
                    groupTitle = groupTitle + titleByKey[group] + ':';
                    titleWithGroup[group] = groupTitle;
                }

                titleWithGroup[s] = groupTitle + titleByKey[s];
            }

            keys = Object.keys(titleByKey);
            keys = keys.sort((x, y) => Q.turkishLocaleCompare(titleWithGroup[x], titleWithGroup[y]));

            return keys;
        }

        get_value(): UserPermissionRow[] {

            let result: UserPermissionRow[] = [];

            for (let item of this.view.getItems()) {
                if (item.GrantRevoke != null && item.Key.charAt(item.Key.length - 1) != ':') {
                    result.push({ PermissionKey: item.Key, Grant: item.GrantRevoke });
                }
            }

            return result;
        }

        set_value(value: UserPermissionRow[]) {

            for (let item of this.view.getItems()) {
                item.GrantRevoke = null;
            }

            if (value != null) {
                for (let row of value) {
                    let r = this.view.getItemById(row.PermissionKey);
                    if (r) {
                        r.GrantRevoke = Q.coalesce(row.Grant, true);
                    }
                }
            }

            this.setItems(this.getItems());
        }

        get_rolePermissions(): string[] {
            return Object.keys(this.rolePermissions);
        }

        set_rolePermissions(value: string[]) {
            this.rolePermissions = {};

            if (value) {
                for (let k of value) {
                    this.rolePermissions[k] = true;
                }
            }

            this.setItems(this.getItems());
        }
    }
}